### PR TITLE
Add `UNIT_NAME_FOR` directive

### PR DESCRIPTION
**Pull request by liam923**
This PR adds a directive to Merlin named `UNIT_NAME_FOR`. It is similar to the directive `UNIT_NAME`, which tells Merlin what the unit name of the file is. However, it takes an additional argument that tells Merlin the name of the file to apply the directive to. For example, the directive `UNIT_NAME_FOR a Mylib__a` tells Merlin that the unit name for files `a.ml` and `a.mli` is `Mylib__a` (contrast this to `UNIT_NAME Mylib__a`). This is necessary when a `.merlin` configuration file is used on a per-directory, as opposed to a per-file, basis.

There currently exists a `WRAPPING_PREFIX` directive, which allows you to specify a prefix to add to all files (ex: `WRAPPING_PREFIX Mylib__`). This is insufficient though, as not all files in a directory may use the same prefix. `WRAPPING_PREFIX` was left for now for compatibility reasons, but I intend to delete it in the near-future.

### CODE REVIEW

File: src/kernel/mconfig.ml [hunk 1]
```diff
@@ -133,7 +134,12 @@ let dump_merlin x =
     );
     "stdlib"       , Json.option Json.string x.stdlib;
     "unit_name"    , Json.option Json.string x.unit_name;
-    "wrapping_prefix" , Json.option Json.string x.wrapping_prefix;
+    "unit_name_for", (let alist =
+                        x.unit_name_for
+                        |> String.Map.map ~f:Json.string
```
Comment (ccasin) [inline src/kernel/mconfig.ml:139]: I don't know what `Json.string` is doing, but it seems this is running it only on the values, not the keys.  Is that what we want?

File: src/kernel/mconfig.ml [hunk 2]
```diff
@@ -266,10 +272,13 @@ let merge_merlin_config dot merlin ~failures ~config_path =
     suffixes = dot.suffixes @ merlin.suffixes;
     stdlib = (if dot.stdlib = None then merlin.stdlib else dot.stdlib);
     unit_name = (if dot.unit_name = None then merlin.unit_name else dot.unit_name);
-    wrapping_prefix =
```
Comment (ccasin) [inline src/kernel/mconfig.ml:269]: Why is `wrapping_prefix` being deleted here?  Probably this belongs  in a future PR that deletes it entirely?

File: src/kernel/mconfig.ml [hunk 3]
```diff
@@ -1027,7 +1037,11 @@ let unitname t =
   | Some name -> Misc.unitname name
   | None ->
     let basename = Misc.unitname t.query.filename in
+    (* CR: get rid of wrapping_prefix. it is only here for legacy reasons at the moment *)
     begin match t.merlin.wrapping_prefix with
     | Some prefix -> prefix ^ basename
-    | None -> basename
+    | None ->
+      String.Map.find_opt basename t.merlin.unit_name_for
+      |> Option.map ~f:Misc.unitname
```
Comment (ccasin) [inline src/kernel/mconfig.ml:1045]: Why is `Misc.unitname` needed here when it wasn't for `wrapping_prefix`?

File: src/kernel/mconfig_dot.ml [hunk 4]
```diff
@@ -283,6 +290,7 @@ let prepend_config ~dir:cwd configurator (directives : directive list) config =
     | `UNKNOWN_TAG tag  ->
       let error =  Printf.sprintf "Unknown configuration tag \"%s\"" tag in
       config, error :: errors
+    | `TAG_ERROR error -> config, error :: errors
```
Comment (ccasin) [inline src/kernel/mconfig_dot.ml:293]: The other error case prints a message - probably we should too?

File: src/kernel/mconfig_dot.ml [hunk 5]
```diff
@@ -300,7 +308,8 @@ let postprocess_config config =
     flags        = clean config.flags;
     stdlib      = config.stdlib;
     unit_name   = config.unit_name;
-    wrapping_prefix = config.wrapping_prefix;
+    unit_name_for = config.unit_name_for;
+    wrapping_prefix = None;
```
Comment (ccasin) [inline src/kernel/mconfig_dot.ml:312]: Probably changes to `wrapping_prefix` should go in their own PR?

File: src/ocaml/typing/env.ml [hunk 6]
```diff
@@ -2881,7 +2881,7 @@ let is_identchar_latin1 = function
   | '\248'..'\255' | '\'' | '0'..'9' -> true
   | _ -> false
 
-let unit_name_of_filename fn =
+let unit_name_for_filename fn =
```
Comment (ccasin) [inline src/ocaml/typing/env.ml:2884]: Why change this?  Looks like an unnecessary diff from the compiler.

File: tests/test-dirs/config/dot-merlin-reader/load-config.t [hunk 7]
```diff
@@ -44,6 +47,10 @@ but we want to observe the stdlib being configured via the STDLIB directive.
     ],
     "stdlib": "/stdlib",
     "unit_name": null,
+    "unit_name_for": {
+      "b": "b",
```
Comment (ccasin) [inline tests/test-dirs/config/dot-merlin-reader/load-config.t:51]: It seems like this should be `Prefix__b` because you set `WRAPPING_PREFIX`?  That seems like a bug - probably we don't care because we aren't going to use `WRAPPING_PREFIX`, but maybe change the test not to use it at all?

**Review by ccasin — COMMENTED** [review]
Various questions/comments below.

File: src/ocaml/typing/env.ml (hunk 6 above)
Comment (liam923) [reply src/ocaml/typing/env.ml:2884]: Whoops that was a stray grep

File: src/kernel/mconfig_dot.ml (hunk 4 above)
Comment (liam923) [reply src/kernel/mconfig_dot.ml:293]: It's a `sprintf` not a `printf` - it's just constructing a string to print later

File: src/kernel/mconfig.ml (hunk 3 above)
Comment (liam923) [reply src/kernel/mconfig.ml:1045]: Ultimately this doesn't matter much - `Misc.unitname` really just capitalized the first letter and drops a file extension if present. Really, I think maybe the `Misc.unitname` should've been included for the wrapping-prefix version.

Comment (liam923) [reply src/kernel/mconfig.ml:1045]: I added the `Misc.unitname` for wrapping prefix

File: src/kernel/mconfig.ml (hunk 1 above)
Comment (liam923) [reply src/kernel/mconfig.ml:139]: `Json.string` = `fun x -> `String x`. This is what we want - an `'Assoc` expects a `(string * Json.t) list`, not a `(Json.t * Json.t) list`.

File: src/kernel/mconfig_dot.ml (hunk 5 above)
Comment (liam923) [reply src/kernel/mconfig_dot.ml:312]: Yep this was a mistake

File: src/kernel/mconfig_dot.ml (hunk 4 above)
Comment (ccasin) [reply src/kernel/mconfig_dot.ml]: Oops - sorry for the noise!

**Review by ccasin — APPROVED** [review]
